k2/csrc/cuda/context.h

```diff
@@ -369,17 +370,22 @@ void Eval(ContextPtrType c, int32_t n, LambdaT &lambda) {
   threads in the same warp will tend to have different values.
   (Of course this doesn't affect the semantics of the operation).
 
+  Note if `lambda` requires additional parameters besides `i` and `j`,
+  user can pass those parameters with `Params... params`,
+  Eval2 will forward them to `lambda`.
+
 */
-template <typename LambdaT>
-void Eval2(cudaStream_t stream, int32_t m, int32_t n, LambdaT &lambda) {
+template <typename LambdaT, typename... Params>
+void Eval2(cudaStream_t stream, int32_t m, int32_t n, LambdaT &lambda,
+           Params... params) {
```

Comment:
`params` here is to accept and forward additional parameters of `lambda`, CPU memory to GPU memory conversion may happen here.